compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -637,24 +644,27 @@ class StructuredOpMatcher {
   /// Adds a predicate that recursively applies other predicates to the
   /// operation defining the `position`-th operand. The position may be
   /// negative, in which case positions are counted from the last one
-  /// Python-style.
+  /// Python-style. When the match is optional, the predicate check succeeds as
+  /// long as the `position` is in bounds. The matcher executed if there is a
+  /// defining operation for the input operand.
   template <typename T>
   std::enable_if_t<
       llvm::is_detected<::mlir::detail::has_operation_or_value_matcher_t, T,
                         Operation *>::value,
       StructuredOpMatcher &>
-  input(int64_t position, T &operandMatcher) {
-    predicates.push_back([position,
-                          &operandMatcher](linalg::LinalgOp linalgOp) -> bool {
+  input(int64_t position, T &operandMatcher,
+        OptionalMatch optional = OptionalMatch(false)) {
+    predicates.push_back([position, &operandMatcher,
+                          optional](linalg::LinalgOp linalgOp) -> bool {
       int64_t transformedPosition =
           position >= 0 ? position : linalgOp.getNumDpsInputs() + position;
       if (transformedPosition >= linalgOp.getNumDpsInputs()) return false;
 
       Operation *definingOp = linalgOp.getDpsInputOperand(transformedPosition)
                                   ->get()
                                   .getDefiningOp();
-      if (!definingOp) return false;
-      return operandMatcher.match(definingOp);
+      if (!definingOp) return optional.value;
+      return operandMatcher.match(definingOp) || optional.value;
```

Comment:
3rd commit: Make match return `false` on `null` and compress this to:
```
return optional.value || operandMatcher.match(definingOp); 
```
?

Reply:
Match is a method of a class template. I prefer guarding here instead of requiring every class passed here to implement a behavior that cannot be checked at compile time using English in documentation. There happens to be a nice demonstration of why comments don't work: the comment above this function says that the matcher should run if there is a defining operation, yet the suggested reversal of the conditions will prevent that from happening.